opm/autodiff/FlowMain.hpp

```diff
@@ -391,6 +391,8 @@ namespace Opm
                 deck_ = parser->parseFile(deck_filename, parseContext);
                 checkDeck(deck_, parser);
                 eclipse_state_.reset(new EclipseState(deck_, parseContext));
+                auto ioConfig = eclipse_state_->getIOConfig();
+                ioConfig->setOutputDir(output_dir_);
```

Comment:
The sequence above looks a bit suspicious: it looks like you extract a local copy of the IOConfig and modify it. After checking the code in opm-parser I see that actually a mutable instance is returned, so we are actually modifying the IOConfig of the `eclipse_state_`. I assume that this is what was intended, but it is not a very intuitive API. If you are meant to be able to change the internal IOConfig of EclipseState that should be done either by putting the required method directly on the EclipseState (i.e. removing the getIOConfig() method and adding a non-const EclipseState::setOutputDir() method) or using an obvious name such as `getMutableIOConfig()` and preferably returning a mutable reference not a shared_prt.

There are some things about const and pointers that are not so intuitive, since the current method actually is const, yet allows you to change the IOConfig member! That is one reason it would be better to return a reference, that would force the method to be mutable, which makes sense.